pkg/epp/controller/inferencemodel_reconciler.go

```diff
@@ -43,44 +47,101 @@ type InferenceModelReconciler struct {
+	logger = logger.WithValues("poolRef", infModel.Spec.PoolRef).WithValues("modelName", infModel.Spec.ModelName)
+	if !c.Datastore.ModelSetIfOlder(infModel) {
+		logger.Info("Skipping InferenceModel, existing instance has older creation timestamp")
+
+	}
+	logger.Info("Added/Updated InferenceModel")
+
 	return ctrl.Result{}, nil
 }
 
-func (c *InferenceModelReconciler) updateDatastore(logger logr.Logger, infModel *v1alpha2.InferenceModel) {
-	loggerDefault := logger.V(logutil.DEFAULT)
+func (c *InferenceModelReconciler) handleModelDeleted(ctx context.Context, req types.NamespacedName) error {
+	logger := log.FromContext(ctx)
+
+	// We will lookup and delete the modelName associated with this object, and search for
+	// other instances referencing the same modelName if exist, and store the oldest in
+	// its place. This ensures that the InferenceModel with the oldest creation
+	// timestamp is active.
+	existing, exists := c.Datastore.ModelDelete(req)
+	if !exists {
+		// No entry exists in the first place, nothing to do.
+		return nil
+	}
+	logger.Info("InferenceModel removed from datastore", "poolRef", existing.Spec.PoolRef, "modelName", existing.Spec.ModelName)
 
-	if infModel.Spec.PoolRef.Name == c.PoolNamespacedName.Name {
-		loggerDefault.Info("Updating datastore", "poolRef", infModel.Spec.PoolRef, "serverPoolName", c.PoolNamespacedName)
-		loggerDefault.Info("Adding/Updating InferenceModel", "modelName", infModel.Spec.ModelName)
-		c.Datastore.ModelSet(infModel)
-		return
+	// List all InferenceModels with a matching ModelName.
+	var models v1alpha2.InferenceModelList
+	if err := c.List(ctx, &models, client.MatchingFields{modelNameKey: existing.Spec.ModelName}, client.InNamespace(c.PoolNamespacedName.Namespace)); err != nil {
+		return fmt.Errorf("listing models that match the modelName %s: %w", existing.Spec.ModelName, err)
+	}
+	if len(models.Items) == 0 {
+		// No other instances of InferenceModels with this ModelName exists.
+		return nil
+	}
+
+	var oldest *v1alpha2.InferenceModel
+	for i := range models.Items {
+		m := &models.Items[i]
+		if m.Spec.ModelName != existing.Spec.ModelName || // The index should filter those out, but just in case!
+			m.Spec.PoolRef.Name != c.PoolNamespacedName.Name || // We don't care about other pools, we could setup an index on this too!
+			m.Name == existing.Name { // We don't care about the same object, it could be in the list if it was only marked for deletion, but not yet deleted.
+			continue
+		}
+		if oldest == nil || m.ObjectMeta.CreationTimestamp.Before(&oldest.ObjectMeta.CreationTimestamp) {
+			oldest = m
+		}
 	}
-	loggerDefault.Info("Removing/Not adding InferenceModel", "modelName", infModel.Spec.ModelName)
-	// If we get here. The model is not relevant to this pool, remove.
-	c.Datastore.ModelDelete(infModel.Spec.ModelName)
+	if oldest != nil && c.Datastore.ModelSetIfOlder(oldest) {
```

Comment:
We should short ciruit ealier in the Reconcile, stop processing if there is already a cached model, and the new one is diff.